pycloudlib/azure/cloud.py

```diff
@@ -0,0 +1,639 @@
+# This file is part of pycloudlib. See LICENSE file for license information.
+"""Azure Cloud type."""
+import base64
+
+import pycloudlib.azure.util as util
+
+from azure.mgmt.resource import ResourceManagementClient
+from azure.mgmt.network import NetworkManagementClient
+from azure.mgmt.compute import ComputeManagementClient
+
+from pycloudlib.cloud import BaseCloud
+from pycloudlib.azure.instance import AzureInstance
+from pycloudlib.key import KeyPair
+from pycloudlib.util import create_tag
+
+
+class Azure(BaseCloud):
+    """Azure Cloud Class."""
+
+    _type = 'azure'
+
+    UBUNTU_RELEASE = {
+        "trusty": "Canonical:UbuntuServer:14.04.0-LTS",
+        "xenial": "Canonical:UbuntuServer:16.04-DAILY-LTS",
+        "bionic": "Canonical:UbuntuServer:18.04-DAILY-LTS",
+        "focal": "Canonical:0001-com-ubuntu-server-focal-daily:20_04-daily-lts"
+    }
+
+    def __init__(self, tag, client_id=None, client_secret_id=None,
+                 subscription_id=None, tenant_id=None):
+        """Initialize the connection to Azure.
+
+        Azure will try to read user credentials form the /home/$USER/.azure
```

Comment:
If client_secret_id, subscription_id or tenant_id are None, let's try to fallback explicitly check for presence of this `~/.azure/credentials.json` config file, raising a RuntimeError if absent with the approriate message that "Azure cloud credentials don't seem to be available. Run: az login or provide client_secret_id and subscription_id to pycloudlib.Azure"

or raise a RuntimeError if not os.path.exists(cred_file = os.path.expanduser('~/.azure/credentials.json')

Reply:
Weird, even being logged into Azure I do not have that file. But since we are calling the `get_client_from_cli_profile`, should we skip this file check and just verify if any of the id parameters are None to raise the exception ?

Reply:
Good pt @lucasmoura. I think it may make sense then to minimally _log.debug("No AZ client  configuration found') when get_client_from_cli_profile raises CLIError

And we should check the known mandatory config_dict values before trying to fallthrough to get_client_from_json_dict
as the errors we get in that case from azure's libs are annoying to decrypt.


Here's a failure I see on a system with no azure cli config
```
csmith@uptown:~/src/pycloudlib$  cat > 1.py <<EOF
import logging
import sys

from pycloudlib import Azure

logging.basicConfig(filename="azure-test.log", level=logging.DEBUG)
azure = Azure(tag='uaclient-integration')
EOF

csmith@uptown:~/src/pycloudlib$  tox -r -e pylint
csmith@uptown:~/src/pycloudlib$ .tox/pylint/bin/python3 ./1.py
csmith@uptown:~/src/pycloudlib$ .tox/pylint/bin/python3 ./1.py 
Traceback (most recent call last):
  File "./1.py", line 7, in <module>
    azure = Azure(tag='uaclient-integration')
  File "/home/csmith/src/pycloudlib/pycloudlib/azure/cloud.py", line 66, in __init__
    self.resource_client = util.get_client(
  File "/home/csmith/src/pycloudlib/pycloudlib/azure/util.py", line 49, in get_client
    client = get_client_from_json_dict(
  File "/home/csmith/src/pycloudlib/.tox/pylint/lib/python3.8/site-packages/azure/common/client_factory.py", line 182, in get_client_from_json_dict
    authority_url = authority_url + '/' + config_dict['tenantId']
TypeError: can only concatenate str (not "NoneType") to str
```



Here's a patch of some of what I was thinking (though you'll see it conflicts with some individual comments I'll have in a minute or two
```diff
diff --git a/pycloudlib/azure/util.py b/pycloudlib/azure/util.py
index 378ca51..e33baf4 100644
--- a/pycloudlib/azure/util.py
+++ b/pycloudlib/azure/util.py
@@ -3,6 +3,9 @@
 from azure.common.client_factory import (get_client_from_cli_profile,
                                          get_client_from_json_dict)
 from knack.util import CLIError
+import logging
+
+logger = logging.getLogger(__name__)
 
 
 def get_client(resource, config_dict):
@@ -24,34 +27,41 @@ def get_client(resource, config_dict):
 
     """
     try:
-        client = get_client_from_cli_profile(resource)
-    except CLIError:
-        client = None
-
-    if client is None:
-        parameters = {
-            "subscriptionId": config_dict.get("subscription_id"),
-            "tenantId": config_dict.get("tenant_id"),
-            "clientId": config_dict.get("client_id"),
-            "clientSecret": config_dict.get("client_secret"),
-            "activeDirectoryEndpointUrl":
-                "https://login.microsoftonline.com",
-            "resourceManagerEndpointUrl": "https://management.azure.com/",
-            "activeDirectoryGraphResourceId":
-                "https://graph.windows.net/",
-            "sqlManagementEndpointUrl":
-                "https://management.core.windows.net:8443/",
-            "galleryEndpointUrl": "https://gallery.azure.com/",
-            "managementEndpointUrl":
-                "https://management.core.windows.net/"
-        }
-
-        client = get_client_from_json_dict(
-            resource,
-            parameters
+        return get_client_from_cli_profile(resource)
+    except Exception:
+        logger.debug(
+            "No valid azure-cli config found. Trying explicit config params"
         )
-
-    return client
+    required_keys = frozenset({"tenant_id"})
+    import pdb; pdb.set_trace()
+    missing_keys = required_keys.difference(set(config_dict.keys()))
+    if missing_keys:
+        raise RuntimeError(
+            "No AZ cli config found, missing required keys: {}".format(
+                ", ".join(missing_keys)
+            )
+        )
+    parameters = {
+        "subscriptionId": config_dict.get("subscription_id"),
+        "tenantId": config_dict.get("tenant_id"),
+        "clientId": config_dict.get("client_id"),
+        "clientSecret": config_dict.get("client_secret"),
+        "activeDirectoryEndpointUrl":
+            "https://login.microsoftonline.com",
+        "resourceManagerEndpointUrl": "https://management.azure.com/",
+        "activeDirectoryGraphResourceId":
+            "https://graph.windows.net/",
+        "sqlManagementEndpointUrl":
+            "https://management.core.windows.net:8443/",
+        "galleryEndpointUrl": "https://gallery.azure.com/",
+        "managementEndpointUrl":
+            "https://management.core.windows.net/"
+    }
+
+    return get_client_from_json_dict(
+        resource,
+        parameters
+    )
 
 
 def get_offer_from_image_id(image_id):
```

Reply:
Yes, that's a great solution. I will add it to the code. Thanks @blackboxsw